File: sdks/go/media_upload.go
package modula

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"mime/multipart"
	"net/http"
)

// MediaUploadResource provides media file upload operations.
type MediaUploadResource struct {
	http *httpClient
}

// MediaUploadOptions configures optional parameters for media uploads.
type MediaUploadOptions struct {
	// Path sets the S3 key path prefix for organizing media files.
	// Segments are separated by "/". Leading and trailing slashes are stripped server-side.
	// Examples: "products/shoes", "blog/headers".
	// When empty, the server defaults to date-based organization (YYYY/M).
	Path string
}

// Upload uploads a file and returns the created media entity.
// An optional MediaUploadOptions can control the storage path.
func (m *MediaUploadResource) Upload(ctx context.Context, r io.Reader, filename string, opts *MediaUploadOptions) (*Media, error) {
	var buf bytes.Buffer
	writer := multipart.NewWriter(&buf)

	part, err := writer.CreateFormFile("file", filename)
	if err != nil {
		return nil, fmt.Errorf("modula: create form file: %w", err)
	}

	if _, err := io.Copy(part, r); err != nil {
		return nil, fmt.Errorf("modula: copy file data: %w", err)
	}

	if opts != nil && opts.Path != "" {
		pathField, err := writer.CreateFormField("path")
		if err != nil {
			return nil, fmt.Errorf("modula: create path field: %w", err)
		}
		if _, err := pathField.Write([]byte(opts.Path)); err != nil {
			return nil, fmt.Errorf("modula: write path field: %w", err)
		}
	}

	if err := writer.Close(); err != nil {
		return nil, fmt.Errorf("modula: close multipart writer: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, m.http.baseURL+"/api/v1/media", &buf)
	if err != nil {
		return nil, fmt.Errorf("modula: create upload request: %w", err)
	}
	req.Header.Set("Content-Type", writer.FormDataContentType())

	resp, err := m.http.doRaw(ctx, req)
	if err != nil {
		return nil, err
	}
	defer func() {
		io.Copy(io.Discard, resp.Body) //nolint: body must be drained
		resp.Body.Close()              //nolint: close on read-only response
	}()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		body, _ := io.ReadAll(resp.Body) //nolint: best-effort read for error message
		return nil, &ApiError{
			StatusCode: resp.StatusCode,
			Body:       string(body),
		}
	}

	var result Media
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("modula: decode upload response: %w", err)
	}
	return &result, nil
}

File: sdks/typescript/modulacms-admin-sdk/src/resources/media-upload.ts
/**
 * Media upload resource for uploading files via multipart/form-data
 * to `POST /media`.
 *
 * @remarks The initial upload response may return `srcset: null` because
 * responsive image variants are generated asynchronously. Poll
 * `client.media.get()` after a delay to retrieve the processed `srcset`.
 *
 * @module resources/media-upload
 * @internal
 */

import type { HttpClient } from '../http.js'
import type { ApiError, RequestOptions } from '../types/common.js'
import { isApiError } from '../types/common.js'
import type { Media } from '../types/media.js'

/**
 * Options for media upload operations.
 */
type MediaUploadOptions = RequestOptions & {
  /**
   * Optional S3 key path prefix for organizing media files.
   *
   * Controls the directory structure in object storage. Segments are
   * separated by `/`. Leading and trailing slashes are stripped server-side.
   *
   * @example `'products/shoes'` — stores as `products/shoes/filename.jpg`
   * @example `'blog/headers'` — stores as `blog/headers/filename.jpg`
   *
   * When omitted, the server defaults to date-based organization (`YYYY/M`).
   */
  path?: string
}

/**
 * Media upload operations available on `client.mediaUpload`.
 */
type MediaUploadResource = {
  /**
   * Upload a file to the CMS media library.
   *
   * Sends a `multipart/form-data` POST request. The `Content-Type` header
   * is set automatically by the browser/runtime (not `application/json`).
   *
   * @param file - The file to upload (browser `File` or `Blob`).
   * @param opts - Optional upload options (abort signal, path).
   * @returns The created media entity. Note: `srcset` may be `null` initially.
   * @throws {@link ApiError} on non-2xx responses.
   * @throws `TypeError` on network failure.
   */
  upload: (file: File | Blob, opts?: MediaUploadOptions) => Promise<Media>
}

/**
 * Create the media upload resource.
 *
 * @param http - Configured HTTP client (used for the `raw` method).
 * @param defaultTimeout - Default timeout in milliseconds.
 * @param credentials - Fetch credentials mode.
 * @param apiKey - Optional API key for the Authorization header.
 * @returns A {@link MediaUploadResource} with an `upload` method.
 * @internal
 */
function createMediaUploadResource(
  http: HttpClient,
  defaultTimeout: number,
  credentials: RequestCredentials,
  apiKey?: string,
): MediaUploadResource {
  return {
    async upload(file: File | Blob, opts?: MediaUploadOptions): Promise<Media> {
      const form = new FormData()
      form.append('file', file)
      if (opts?.path !== undefined) {
        form.append('path', opts.path)
      }

      const headers: Record<string, string> = {}
      if (apiKey) {
        headers['Authorization'] = `Bearer ${apiKey}`
      }

      let signal: AbortSignal
      if (opts?.signal) {
        const controller = new AbortController()
        const timeoutSignal = AbortSignal.timeout(defaultTimeout)

        const onAbort = (source: AbortSignal) => () => {
          if (!controller.signal.aborted) controller.abort(source.reason)
        }

        timeoutSignal.addEventListener('abort', onAbort(timeoutSignal), { once: true })
        opts.signal.addEventListener('abort', onAbort(opts.signal), { once: true })

        if (opts.signal.aborted) controller.abort(opts.signal.reason)

        signal = controller.signal
      } else {
        signal = AbortSignal.timeout(defaultTimeout)
      }

      const response = await http.raw('/media', {
        method: 'POST',
        headers,
        credentials,
        signal,
        body: form,
      })

      if (!response.ok) {
        const ct = response.headers.get('content-type')
        const isJson = ct !== null && ct.includes('application/json')
        let body: unknown
        let message = response.statusText
        if (isJson) {
          body = await response.json()
        } else {
          const text = await response.text()
          if (text) {
            message = text.trim()
            body = text.trim()
          }
        }
        const err: ApiError = {
          _tag: 'ApiError' as const,
          status: response.status,
          message,
          body,
        }
        throw err
      }

      return response.json() as Promise<Media>
    },
  }
}

/**
 * Check if an error is a duplicate media conflict (HTTP 409).
 * Thrown when uploading a file that already exists in the media library.
 */
function isDuplicateMedia(err: unknown): err is ApiError {
  return isApiError(err) && err.status === 409
}

/**
 * Check if an error is a file-too-large rejection (HTTP 400).
 * Thrown when the uploaded file exceeds the server's size limit.
 */
function isFileTooLarge(err: unknown): err is ApiError {
  return isApiError(err) && err.status === 400 && typeof err.message === 'string' && err.message.toLowerCase().includes('too large')
}

/**
 * Check if an error is an invalid media path rejection (HTTP 400).
 * Thrown when the `path` option contains invalid characters or path traversal.
 */
function isInvalidMediaPath(err: unknown): err is ApiError {
  return isApiError(err) && err.status === 400 && typeof err.message === 'string' && (err.message.includes('path traversal') || err.message.includes('invalid character in path'))
}

export type { MediaUploadResource, MediaUploadOptions }
export { createMediaUploadResource, isDuplicateMedia, isFileTooLarge, isInvalidMediaPath }
